src/rgw/rgw_redis_scripts.lua

```diff
@@ -0,0 +1,231 @@
+    if redis.call('EXISTS', name) == 1 then
+        local existing_cookie = redis.call('GET', name)
+        if existing_cookie == cookie then
+            return formatResponse(0, "", "")
+        else
+            return formatResponse(-lerrorCodes.EBUSY, "Lock is held by another process", "")
+        end
+    end
+    return formatResponse(-lerrorCodes.ENOENT, "Lock does not exist", "")
+end
+
+--- Acquire a lock on a resource.
+--- It sets a key with a cookie value if the key does not exist.
+--- If the key exists and the value is same as cookie, it extends the lock.
+--- If the key exists and the value is different from cookie, it fails.  
+---@param keys table A single element list - lock name
+---@param args table A two-element list - cookie and timeout
+---@return number 0 if the lock is acquired or extended
+local function lock(keys, args)
+    local name = keys[1]
+    local cookie = args[1]
+    local timeout = args[2]
+    local lock_status = assert_lock(keys, args)
+    if lock_status.map.errorCode == 0 then
+        redis.call('PEXPIRE', name, timeout)
+        return formatResponse(0, "", "")
+    elseif lock_status.map.errorCode == -lerrorCodes.ENOENT then
+        redis.call('SET', name, cookie, 'PX', timeout)
+        return formatResponse(0, "", "")
+    end
+    return lock_status
+end
+
+--- Release the lock on a resource.
+--- It deletes the key if the value matches the cookie.
+---@param keys table A single element list - lock name
+---@param args table A single-element list - cookie
+local function unlock(keys, args)
+    local name = keys[1]
+    local cookie = args[1]
+    local lock_status = assert_lock(keys, args)
+    if lock_status.map.errorCode == 0 then
+        redis.call('DEL', name)
+        return formatResponse(0, "", "")
+    end
+    return lock_status
+end
+
+--- Register the functions.
+redis.register_function('lock', lock)
+redis.register_function('unlock', unlock)
+redis.register_function('assert_lock', assert_lock)
+
+-----
+--- Queue functions
+-----
+
+-- Refer to 'Execution under low memory conditions'
+-- in https://redis.io/docs/latest/develop/interact/programmability/eval-intro/
+---
+
+-- Function to generate a string longer than n by concatenating the charset
+local function generateString(n)
+    local charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+    local result = {}
+    local charset_length = #charset
+
+    local repeats = math.ceil(n / charset_length) + 1
+    for i = 1, repeats do
+        table.insert(result, charset)
+    end
+    return table.concat(result)
+end
+
+--- Add an item to the reserve queue for n bytes
+--- @param keys table A single element list - queue name
+--- @param args table A single-element list - item size
+--- @return number 0 if the item is added to the queue, -lerrorCodes.ENOMEM if the queue is full
+local function reserve(keys, args)
+    local name = "reserve:" .. keys[1]
+    local item_size = tonumber(args[1])
```

Comment:
what if `args[1]` is not a valid number?

Reply:
Redis returns an error. 
For us, it should always be, that is a  requirement of the call.

Reply:
you mean that redis checks that and reply with an error without invoking the function?

Reply:
No, the script fails with 
```
ERR user_function:107: attempt to perform arithmetic on local 'n' (a nil value) script: reserve, on @user_function:107.
```

Reply:
so, please add more explicit error handling in the lua script.
e.g.
```lua
local item_size = tonumber(args[1])
if item_size == nil
  -- return an error

```
there are probably other places like that